packages/crypto-messages/source/proposal.ts

```diff
@@ -1,80 +1,107 @@
-import { Contracts } from "@mainsail/contracts";
+import { inject, injectable } from "@mainsail/container";
+import { Contracts, Identifiers } from "@mainsail/contracts";
 
+@injectable()
 export class Proposal implements Contracts.Crypto.Proposal {
-	#round: number;
+	@inject(Identifiers.Cryptography.Message.Factory)
+	private readonly messageFactory!: Contracts.Crypto.MessageFactory;
+
+	#height!: number;
+	#round!: number;
 	#validRound?: number;
-	#block: Contracts.Crypto.ProposedBlock;
-	#validatorIndex: number;
-	#signature: string;
-	#serialized: Buffer;
+	#dataSerialized!: string;
+	#data?: Contracts.Crypto.ProposedData;
+	#validatorIndex!: number;
+	#signature!: string;
+	#serialized!: Buffer;
 
-	constructor({
+	public initialize({
 		round,
 		validatorIndex,
-		block,
+		height,
+		dataSerialized,
 		validRound,
 		signature,
 		serialized,
-	}: Contracts.Crypto.ProposalData & { block: Contracts.Crypto.ProposedBlock; serialized: Buffer }) {
+	}: Contracts.Crypto.ProposalData & { dataSerialized: string; height: number; serialized: Buffer }): Proposal {
+		this.#height = height;
 		this.#round = round;
 		this.#validRound = validRound;
-		this.#block = block;
+		this.#dataSerialized = dataSerialized;
 		this.#validatorIndex = validatorIndex;
 		this.#signature = signature;
 		this.#serialized = serialized;
+
+		return this;
 	}
 
-	get height(): number {
-		return this.#block.block.header.height;
+	public get isDataDeserialized(): boolean {
+		return this.#data !== undefined;
 	}
 
-	get round(): number {
-		return this.#round;
+	public get height(): number {
+		return this.#height;
 	}
 
-	get validRound(): number | undefined {
-		return this.#validRound;
+	public get round(): number {
+		return this.#round;
 	}
 
-	get block(): Contracts.Crypto.ProposedBlock {
-		return this.#block;
+	public get validRound(): number | undefined {
+		return this.#validRound;
 	}
 
-	get validatorIndex(): number {
+	public get validatorIndex(): number {
 		return this.#validatorIndex;
 	}
 
-	get signature(): string {
+	public get signature(): string {
 		return this.#signature;
 	}
 
-	get serialized(): Buffer {
+	public get serialized(): Buffer {
 		return this.#serialized;
 	}
 
-	toString(): string {
+	public async deserializeData(): Promise<void> {
+		if (this.#data !== undefined) {
+			return;
+		}
+
+		this.#data = await this.messageFactory.makeProposedDataFromBytes(Buffer.from(this.#dataSerialized, "hex"));
```

Comment:
We deserialize the header here https://github.com/ArkEcosystem/mainsail/pull/551/files#diff-124e70168ef6848b8219ef87e60cf96fb3b4f218c6355ae104e1ab437d5ed430R166

So we also should be able to skip the header deserialization inside `deserializeWithTransactions` and just look at the payload unless you foresee any issues

Reply:
I will address this in the new PR if necessary. Header deserialization shouldn't take so long. The most consuming part is deserializing big block payload. We have 2 options:
- deserialize only height in the proposal
- deserialize header then block payload as you suggested